### PR TITLE
planner: make pattern match case-insensitive for some infoschema tables

### LINKED ISSUE #56377: `information_schema.tables` return wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
use test;
drop table if exists caseSensitive;
drop table if exists unrelatedTable;

create table caseSensitive (a int);
create table unrelatedTable (a int);
select table_schema, table_name from information_schema.tables where table_schema = 'test' and table_name like '%aseSensitive';
select table_schema, table_name, tidb_pk_type from information_schema.tables where table_schema = 'test' and table_name like '%aseSensitive';
```

### 2. What did you expect to see? (Required)
```sql
mysql> select table_schema, table_name from information_schema.tables where table_schema = 'test' and table_name like '%aseSensitive';
+--------------+---------------+
| table_schema | table_name    |
+--------------+---------------+
| test         | caseSensitive |
+--------------+---------------+
1 row in set (0.00 sec)

mysql> select table_schema, table_name, tidb_pk_type from information_schema.tables where table_schema = 'test' and table_name like '%aseSensitive';
+--------------+---------------+--------------+
| table_schema | table_name    | tidb_pk_type |
+--------------+---------------+--------------+
| test         | caseSensitive | NONCLUSTERED |
+--------------+---------------+--------------+
1 row in set (0.00 sec)
```

### 3. What did you see instead (Required)

```sql
mysql> select table_schema, table_name from information_schema.tables where table_schema = 'test' and table_name like '%aseSensitive';
+--------------+----------------+
| table_schema | table_name     |
+--------------+----------------+
| test         | unrelatedTable |
| test         | caseSensitive  |
+--------------+----------------+
2 rows in set (0.00 sec)

mysql> select table_schema, table_name, tidb_pk_type from information_schema.tables where table_schema = 'test' and table_name like '%aseSensitive';
Empty set (0.00 sec)
```

### 4. What is your TiDB version? (Required)

b520f61d9f

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #56377

Problem Summary: 

There are two problems with #55844:
  - When pattern match is pushed down to memory table reader, it is case sensitive. For example, `LIKE '%T'` cannot match `"tt"`.
  - Pattern match does not work for the fast path introduced by #55574.

### What changed and how does it work?

- Make memory table reader pattern match case-insensitive.
- Consider pattern match when filtering schema object.
- Sort the result returned from `ListSchemasAndTables` to make test stable.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 1]
```diff
@@ -661,6 +672,13 @@ func findTableAndSchemaByName(
 		schemaSlice = append(schemaSlice, st.schema)
 		tableSlice = append(tableSlice, st.table)
 	}
+	sort.Slice(schemaSlice, func(i, j int) bool {
```
Comment (joechenrh) [inline pkg/planner/core/memtable_infoschema_extractor.go:675]: We only need to sort by schema name?

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 2]
```diff
@@ -721,6 +739,13 @@ func findSchemasForTables(
 			remains = append(remains, tbl)
 		}
 	}
+	sort.Slice(schemaSlice, func(i, j int) bool {
```
Comment (joechenrh) [inline pkg/planner/core/memtable_infoschema_extractor.go:742]: Can we create a new slice to store tables instead of directly modifing input slice?

**Review by tiancaiamao — APPROVED** [review]

File: pkg/planner/core/memtable_infoschema_extractor.go (hunk 2 above)
Comment (tangenta) [reply pkg/planner/core/memtable_infoschema_extractor.go:742]: The input slice is not used by anyone else. Let me add some comments for this.

File: pkg/planner/core/memtable_infoschema_extractor.go (hunk 1 above)
Comment (tangenta) [reply pkg/planner/core/memtable_infoschema_extractor.go:675]: https://github.com/pingcap/tidb/pull/56378/commits/291d8cc58a967e9e2f13c710f2573ec712fe4b1c

File: pkg/planner/core/memtable_infoschema_extractor.go (hunk 2 above)
Comment (tangenta) [reply pkg/planner/core/memtable_infoschema_extractor.go:747]: https://github.com/pingcap/tidb/pull/56378/commits/291d8cc58a967e9e2f13c710f2573ec712fe4b1c

**Review by joechenrh — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**